crates/ruff_python_parser/src/lexer.rs

```diff
@@ -579,6 +718,19 @@ impl<'source> Lexer<'source> {
     // This is the main entry point. Call this function to retrieve the next token.
     // This function is used by the iterator implementation.
     pub fn next_token(&mut self) -> LexResult {
+        if let Some(fstring_context) = self.fstring_stack.last() {
+            if !fstring_context.is_in_expression() {
+                self.cursor.start_token();
+                if let Some(tok) = self.maybe_lex_fstring_middle()? {
+                    return Ok((tok, self.token_range()));
+                }
+                if let Some(tok) = self.maybe_lex_fstring_end() {
+                    self.fstring_stack.pop();
+                    return Ok((tok, self.token_range()));
+                }
+            }
```

Comment:
Nit: I wonder if it would make sense to refactor `next_token` a bit so that we could call into a `lex_fstring_expression` that:

* Doesn't handle indentation
* overrides the lexing for fstring- specific tokens 

But I'm not sure if its worth the complexity (and work)